cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/servlet/security/jwt/JwtSecurityProviderIntegrationTest.java

```diff
@@ -0,0 +1,192 @@
+      }
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+      super.doStart();
+      _loginService.start();
+    }
+
+    @Override
+    protected void doStop() throws Exception {
+      _loginService.stop();
+      super.doStop();
+    }
+  }
+
+  public JwtSecurityProviderIntegrationTest() throws Exception {
+    _tokenAndKeys = TokenGenerator.generateToken(TEST_USERNAME);
+    _publicKeyFile = createCertificate(_tokenAndKeys);
+    _tokenProviderServer = new Server(0);
+    _tokenProviderServer.setHandler(new TestAuthenticatorHandler());
+  }
+
+  /**
+   * Starts the token provider and the Kafka environment before test
+   * @throws Exception
+   */
+  @Before
+  public void setup() throws Exception {
+    _tokenProviderServer.start();
+    super.start();
+  }
+
+  /**
+   * Stops the token provider and the Kafka environment after test
+   * @throws Exception
+   */
+  @After
+  public void teardown() throws Exception {
+    super.stop();
+    _tokenProviderServer.stop();
+  }
+
+  @Override
+  protected Map<String, Object> withConfigs() {
+    Map<String, Object> securityConfigs = new HashMap<>();
+    securityConfigs.put(WebServerConfig.WEBSERVER_SECURITY_ENABLE_CONFIG, true);
+    securityConfigs.put(WebServerConfig.WEBSERVER_SECURITY_PROVIDER_CONFIG, JwtSecurityProvider.class);
+    securityConfigs.put(WebServerConfig.JWT_AUTHENTICATION_PROVIDER_URL_CONFIG,
+        _tokenProviderServer.getURI().toString() + "?" +
+            TEST_USERNAME_KEY + "=" + TEST_USERNAME +
+            "&" + TEST_PASSWORD_KEY + "=" + TEST_PASSWORD +
+            "&origin={redirectUrl}");
+    securityConfigs.put(WebServerConfig.WEBSERVER_AUTH_CREDENTIALS_FILE_CONFIG,
+        Objects.requireNonNull(this.getClass().getClassLoader().getResource("jwt-auth.credentials")).getPath());
+    securityConfigs.put(WebServerConfig.JWT_COOKIE_NAME_CONFIG, JWT_TOKEN_COOKIE_NAME);
+    securityConfigs.put(WebServerConfig.JWT_AUTH_CERTIFICATE_LOCATION_CONFIG, _publicKeyFile.getAbsolutePath());
+
+    return securityConfigs;
+  }
+
+  @Test
+  public void testSuccessfulLogin() throws Exception {
+    HttpURLConnection stateEndpointConnection = (HttpURLConnection) new URI(_app.serverUrl())
+        .resolve(CRUISE_CONTROL_STATE_ENDPOINT).toURL().openConnection();
+    String cookie = stateEndpointConnection.getHeaderField(HttpHeader.SET_COOKIE.asString());
+    stateEndpointConnection = (HttpURLConnection) new URI(_app.serverUrl())
+        .resolve(CRUISE_CONTROL_STATE_ENDPOINT).toURL().openConnection();
+    stateEndpointConnection.setRequestProperty(HttpHeader.COOKIE.asString(), cookie);
+    assertEquals(HttpServletResponse.SC_OK, stateEndpointConnection.getResponseCode());
+  }
+
+  private File createCertificate(TokenGenerator.TokenAndKeys tokenAndKeys) throws Exception {
+    String subjectDN = "C=US, ST=California, L=Santa Clara, O=LinkedIn, CN=localhost";
+    Provider bcProvider = new BouncyCastleProvider();
```

Comment:
I am curious why BouncyCastle was used. Was there limitation in generating self-signed certificates with classes provided with JDK (for example CertAndKeyGen class)?

Reply:
There aren't really good classes in plain Java to work with on this level (as far as I saw). `CertAndKeyGen` is unfortunately a Sun package and therefore belongs to Oracle so I think we should avoid it as it isn't compatible with OpenJDK.
Additionally to this, BouncyCastle was already a dependency of the metrics reporter tests so I thought it's not a too big leap to use this here.

Reply:
Ah ok, understood.